### PR TITLE
Optimized the multithreading strategy for -listxml

### PR DESCRIPTION
**Pull request by npwoods**
The previous algorithm would spawn tasks but wait for them in sequential order.  This is not necessarily optimal, and with these changes we will respond to which ever task may complete first.

On my computer (Quad Core 2), this triples the speed of a full -listxml

### CODE REVIEW

**Comment by cuavas** [comment]
Does this cause the output to change non-deterministically depending on timing?  I don’t want a situation where the same binary can produce different `-listxml` output situationally.

**Comment by npwoods** [comment]
Good call; you're right the previous changes made the results not be deterministic

I've changed around the logic and got comparable speed increases while ensuring deterministic results

File: src/frontend/mame/infoxml.cpp [hunk 1]
```diff
@@ -404,79 +455,97 @@ void info_xml_creator::output(std::ostream &out, const std::function<bool(const
...
-			if (!drivlist.next())
-			{
-				// at this point we are done enumerating through drivlist and it is no
-				// longer safe to call next(), so record that we're done
-				drivlist_done = true;
-			}
-			else if (!filter || filter(drivlist.driver().name, filter_done))
+			// we want to launch a task; grab a packet of drivers to process
+			std::vector<std::reference_wrapper<const game_driver>> drivers = filtered_drivlist.next(20);
+			if (drivers.size() <= 0)
```
Comment (cuavas) [inline src/frontend/mame/infoxml.cpp:484]: You should use `.empty()` to test for an empty standard container.

**Comment by MASHinfo** [comment]
@ Nathan Woods

> I just did a simple
> mame -listxml >1.xml
> mame -listxml >2.xml
> diff 1.xml 2.xml
> And did not see any differences. Is anybody else seeing this?
The difference is still happend on my Windows XP 32bit MAME Arcade Build.
Your optimized listinfo output did work until vaz softwarelist change?
I think it is only a 32bit problem! Not worth talking about ;)

**Pull request merged**